Tests require CPython C headers which might be omitted by uv

tl;dr: By default, uv will use the python installed on the OS if one compatible is already present, however this Python might not have the C headers installed (e.g. on ubuntu `python3.12` installed but not `python3.12-dev`)

The solution to avoid this is to run `UV_MANAGED_PYTHON=1 uv sync` instead of `uv sync` (which force uv to only use its own Python distribution), I guess the README.md should be updated to reflect this (though adding this exotic environ variable might seems weird, so a longer explanation might be needed there...)

To reproduce:

```
$ apt search ^python3.12
Sorting... Done

Full Text Search... Done
python3.12/noble-security,noble-updates,now 3.12.3-1ubuntu0.15 amd64 [installed,automatic]  <================ Python 3.12 is installed...
  Interactive high-level object-oriented language (version 3.12)
...
python3.12-dev/noble-security,noble-updates 3.12.3-1ubuntu0.15 amd64  <================= ...but not its headers !
  Header files and a static library for Python (v3.12)
...
$ uv sync
...
$ ./.venv/bin/pytest 'spy/tests/compiler/test_py_cffi.py::TestPyCFFI::test_simple[py-cffi]'
================================================================================================ test session starts ================================================================================================
platform linux -- Python 3.12.3, pytest-8.3.4, pluggy-1.6.0
rootdir: /home/emmanuel/projects/spy
configfile: pyproject.toml
plugins: hypothesis-6.155.7, asyncio-1.3.0, pyodide-0.59.2, xdist-3.8.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=function, asyncio_default_test_loop_scope=function
collected 1 item                                                                                                                                                                                                    

spy/tests/compiler/test_py_cffi.py F                                                                                                                                                                          [100%]

===================================================================================================== FAILURES ======================================================================================================
__________________________________________________________________________________________ TestPyCFFI.test_simple[py-cffi] __________________________________________________________________________________________

self = <spy.tests.compiler.test_py_cffi.TestPyCFFI object at 0x73a2b9614080>

    def test_simple(self):
>       mod = self.compile("""
        def add(x: i32, y: i32) -> i32:
            return x + y
        """)

spy/tests/compiler/test_py_cffi.py:7: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
spy/tests/support.py:264: in compile
    outfile = backend.build()  # e.g. 'test.wasm' or 'test.mjs'
spy/backend/c/cbackend.py:232: in build
    return cffi_build(self.build_script)
spy/build/cffi.py:30: in cffi_build
    proc = robust_run(cmdline)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cmdline = ['/home/emmanuel/projects/spy/.venv/bin/python3', '/tmp/pytest-of-emmanuel/pytest-12/test_simple_py_cffi_0/build/cffi/_test-cffi-build.py'], unbuffer = False

    def robust_run(
        cmdline: Sequence[str | py.path.local], unbuffer: bool = False
    ) -> subprocess.CompletedProcess:
        """
        Similar to subprocess.run, but raise an Exception with the content of
        stdout+stderr in case of failure.
    
        If unbuffer is True, the command is run unbuffered using pexpect.
        """
        cmdline_s = [str(x) for x in cmdline]
        if unbuffer:
            # Note that unbuffer doesn't read from stdin by default
            proc = unbuffer_run(cmdline_s)
        else:
            # Use capture_output=True to capture stdout and stderr separately
            proc = subprocess.run(cmdline_s, capture_output=True)
    
        if proc.returncode != 0:
            FORCE_COLORS = True
            lines = ["subprocess failed:"]
            lines.append(" ".join(cmdline_s))
            lines.append("")
            errlines = []
            if proc.stdout:
                errlines += proc.stdout.decode("utf-8").splitlines()
            if proc.stderr:
                errlines += proc.stderr.decode("utf-8").splitlines()
            if FORCE_COLORS:
                errlines = [Color.set("default", line) for line in errlines]
            lines += errlines
            msg = "\n".join(lines)
>           raise Exception(msg)
E           Exception: subprocess failed:
E           /home/emmanuel/projects/spy/.venv/bin/python3 /tmp/pytest-of-emmanuel/pytest-12/test_simple_py_cffi_0/build/cffi/_test-cffi-build.py
E           
E           _test.c:50:14: fatal error: pyconfig.h: No such file or directory
E              50 | #    include <pyconfig.h>
E                 |              ^~~~~~~~~~~~
E           compilation terminated.
E           Traceback (most recent call last):
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/_distutils/spawn.py", line 83, in spawn
E               subprocess.check_call(cmd, env=_inject_macos_ver(env))
E             File "/usr/lib/python3.12/subprocess.py", line 413, in check_call
E               raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '['/usr/bin/x86_64-linux-gnu-gcc', '-fno-strict-overflow', '-Wsign-compare', '-DNDEBUG', '-g', '-O2', '-Wall', '-fPIC', '-I/home/emmanuel/projects/spy/.venv/include', '-I/usr/include/python3.12', '-c', '_test.c', '-o', './_test.o', '-fvisibility=hidden', '-fPIC', '-DSPY_GC_NONE', '-DSPY_TARGET_NATIVE', '-DSPY_DEBUG', '-O0', '-g', '-Werror=implicit-function-declaration', '-I', '/home/emmanuel/projects/spy/spy/libspy/include', '--std=c99', '-Wfatal-errors', '-fdiagnostics-color=always', '-I/tmp/pytest-of-emmanuel/pytest-12/test_simple_py_cffi_0/build/src']' returned non-zero exit status 1.
E           
E           The above exception was the direct cause of the following exception:
E           
E           Traceback (most recent call last):
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/_distutils/compilers/C/unix.py", line 221, in _compile
E               self.spawn(compiler_so + cc_args + [src, '-o', obj] + extra_postargs)
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/_distutils/compilers/C/base.py", line 1151, in spawn
E               spawn(cmd, **kwargs)
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/_distutils/spawn.py", line 89, in spawn
E               raise DistutilsExecError(
E           distutils.errors.DistutilsExecError: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
E           
E           During handling of the above exception, another exception occurred:
E           
E           Traceback (most recent call last):
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/cffi/ffiplatform.py", line 48, in _build
E               dist.run_command('build_ext')
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/dist.py", line 1107, in run_command
E               super().run_command(command)
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 1019, in run_command
E               cmd_obj.run()
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/command/build_ext.py", line 97, in run
E               _build_ext.run(self)
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/_distutils/command/build_ext.py", line 367, in run
E               self.build_extensions()
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/_distutils/command/build_ext.py", line 483, in build_extensions
E               self._build_extensions_serial()
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/_distutils/command/build_ext.py", line 509, in _build_extensions_serial
E               self.build_extension(ext)
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/command/build_ext.py", line 262, in build_extension
E               _build_ext.build_extension(self, ext)
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/_distutils/command/build_ext.py", line 564, in build_extension
E               objects = self.compiler.compile(
E                         ^^^^^^^^^^^^^^^^^^^^^^
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/_distutils/compilers/C/base.py", line 648, in compile
E               self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/setuptools/_distutils/compilers/C/unix.py", line 223, in _compile
E               raise CompileError(msg)
E           distutils.compilers.C.errors.CompileError: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
E           
E           During handling of the above exception, another exception occurred:
E           
E           Traceback (most recent call last):
E             File "/tmp/pytest-of-emmanuel/pytest-12/test_simple_py_cffi_0/build/cffi/_test-cffi-build.py", line 23, in <module>
E               sofile = ffibuilder.compile(verbose=False)
E                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/cffi/api.py", line 727, in compile
E               return recompile(self, module_name, source, tmpdir=tmpdir,
E                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/cffi/recompiler.py", line 1581, in recompile
E               outputfilename = ffiplatform.compile('.', ext,
E                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/cffi/ffiplatform.py", line 20, in compile
E               outputfilename = _build(tmpdir, ext, compiler_verbose, debug)
E                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E             File "/home/emmanuel/projects/spy/.venv/lib/python3.12/site-packages/cffi/ffiplatform.py", line 54, in _build
E               raise VerificationError('%s: %s' % (e.__class__.__name__, e))
E           cffi.VerificationError: CompileError: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1

spy/util.py:205: Exception
============================================================================================== short test summary info ==============================================================================================
FAILED spy/tests/compiler/test_py_cffi.py::TestPyCFFI::test_simple[py-cffi] - Exception: subprocess failed:
================================================================================================= 1 failed in 0.64s =================================================================================================
```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -61,6 +61,14 @@ spy = "spy.cli:pyproject_entry_point"
 requires = ["setuptools"]
 build-backend = "setuptools.build_meta"
 
+[tool.uv]
+# Instruct uv to never use the system Python installations.
+# This is to avoid errors during CFFI tests (i.e. when building
+# a native extension) when the system Python doesn't contain the
+# Python C headers (e.g. `python3-dev` package not installed).
+# See: https://github.com/spylang/spy/issues/621
+python-preference = "only-managed"
+
 [tool.setuptools]
 include-package-data = true
 

